Fixed implementation of BOP (#1402)

Setu-Gupta opened the pull request:
This PR fixes the issues with the implementation of the Best Offset Prefetcher described in issue #1402

On branch bop
Changes to be committed:
	modified:   src/mem/cache/prefetch/bop.cc
	modified:   src/mem/cache/prefetch/bop.hh

Commit:
Fixed implementation of BOP

This commit fixes the issues with the implementation of the Best Offset
Prefetcher.

GitHub Issue: https://github.com/gem5/gem5/issues/1402

On branch bop
Changes to be committed:
	modified:   src/mem/cache/prefetch/bop.cc
	modified:   src/mem/cache/prefetch/bop.hh

Co-authored-by: Abhishek Shailendra Singh <abs218@leigh.edu>
Change-Id: Iaa111ebe1e903787af961b50d9b5d2b44aabb755

ivanaamit commented:
Hi @Setu-Gupta,

None of the commits in this pull request contains a Change-ID, which we require for any changes made to gem5.  To automatically insert one, run the following:

```
f=.git/hooks/commit-msg
mkdir -p $f
curl -Lo $f https://gerrit-review.googlesource.com/tools/hooks/commit-msg
chmod +x $f
```

Then, amend the commit with `git commit --amend --no-edit`, and update your pull request. Thank you.

Additionally, we have a specific style that every commit message must follow. You can find the guidelines here: https://www.gem5.org/contributing (search for "Committing").

Setu-Gupta commented:
Hi @ivanaamit 

Thanks for letting me know to fix the PR. I've amended the commit message. I've also added the information of the co-author (@AbhishekUoR) of this commit.

Thanks

Commit:
mem: Fixed Prefetcher.py

Added a paramter for the prefetch degree of Best Offset Prefetcher

On branch bop
Your branch is up to date with 'origin/bop'.

Changes to be committed:
	modified:   src/mem/cache/prefetch/Prefetcher.py

Co-authored-by: Abhishek Shailendra Singh <abs218@leigh.edu>
Change-Id: Ic2039b549dde9a59afb32a8066e8abb7c9ffbff9

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -47,18 +47,14 @@ BOP::BOP(const BOPPrefetcherParams &p)
-        fatal("%s: negative offsets enabled with odd offset list size\n",
-              name());
-    }
 
```
Add fatal_if(degree <= 0, ...)

Setu-Gupta replied:
Fixed in the latest commit. I also fixed the check that I had removed.

odanrc commented on src/mem/cache/prefetch/bop.hh:
```diff
@@ -151,6 +153,7 @@ class BOP : public Queued
         void notifyFill(const CacheAccessProbeArg &arg) override;
 
     public:
+    	unsigned int degree;
```
Add doxygen documentation

Setu-Gupta replied:
Fixed in the latest commit

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -114,22 +111,24 @@ BOP::delayQueueEventWrapper()
-    Addr hash1 = addr >> way;
-    Addr hash2 = hash1 >> floorLog2(rrEntries);
-    return (hash1 ^ hash2) & (Addr)(rrEntries - 1);
+    Addr lrrEntries = floorLog2(rrEntries);
```
Main: Change lrrEntries -> lrr_entries
Nit: Change lrr_entries to log_rr_entries, because this significantly improves code readability (I only understood what ´l´ meant when I saw the fix in the tag function).

From gem5´s coding style (https://www.gem5.org/documentation/general_docs/development/coding_style/):
Local variables are lower case, with underscores separating words (e.g., local_variable).

Setu-Gupta replied:
Fixed in the latest commit

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -114,22 +111,24 @@ BOP::delayQueueEventWrapper()
-    Addr hash2 = hash1 >> floorLog2(rrEntries);
-    return (hash1 ^ hash2) & (Addr)(rrEntries - 1);
+    Addr lrrEntries = floorLog2(rrEntries);
+    Addr lineaddr = addr >> lBlkSize;
```
Ditto line_addr

Setu-Gupta replied:
Fixed in the latest commit

odanrc replied:
This one was not updated either

Setu-Gupta replied:
Fixed now

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -201,27 +207,34 @@ BOP::bestOffsetLearning(Addr x)
-            bestOffset = phaseBestOffset;
+    // Check if its time to re-calculate the best offset
+    if ((bestScore >= scoreMax) || (round >= roundMax))
+    {
```
Check the Braces section in the coding style:
For control blocks (if, while, etc.), opening braces must be on the same line as the control keyword with a space between the closing parenthesis and the opening brace.

Setu-Gupta replied:
Fixed in the latest commit

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -114,22 +111,24 @@ BOP::delayQueueEventWrapper()
 
 unsigned int
-BOP::hash(Addr addr, unsigned int way) const
+BOP::index(Addr addr, unsigned int way) const
```
Nit: in general it is better to avoid these modifications that don´t really change anything, as this could needlessly create merge conflicts for other users. There is no need to revert, though, as it is quite unlikely that someone is modifying the prefetcher.

Setu-Gupta replied:
The reason why I changed the function name to index is because the original author's code which was submitted for DPC-2 used the word index instead of hash. Also, while I was reading this code, the first time I came across this function, the name `hash` confused me as this function was actually being used to index the RR Table. That's why I thought it would be better to change the name to something more meaningful

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -185,14 +184,21 @@ BOP::testRR(Addr addr) const
 
 void
-BOP::bestOffsetLearning(Addr x)
+BOP::bestOffsetLearning(Addr addr)
```
Since addr is not an address but a tag, rename it to addr_tag or something similar. Then update the comment below to remove the outdated information

Setu-Gupta replied:
Fixed in the latest commit

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -201,27 +207,34 @@ BOP::bestOffsetLearning(Addr x)
+                resetScores();
+                issuePrefetchRequests = true;
+            }
+            else
```
nit: I suggest adding braces to the else too, in order to make it less easy to add bugs in the future

Setu-Gupta replied:
Fixed in the latest commit

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -201,27 +207,34 @@ BOP::bestOffsetLearning(Addr x)
+            if (bestScore > badScore)
+            {
+                bestOffset = phaseBestOffset;
+                round = 0;
```
We already assign 0 to round, bestScore and phaseBestOffset, and call resetScores outside this conditional block, so something is looking fishy here. Is it a bug or is it duplicated code?

Setu-Gupta replied:
Fixed in the latest commit

odanrc replied:
It does not seem to have changed in the latest commit

Setu-Gupta replied:
Fixed now. It was indeed duplicate code. Thanks for pointing it out. I've removed the redundant code

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -234,21 +247,21 @@ BOP::calculatePrefetch(const PrefetchInfo &pfi,
-        Addr prefetch_addr = addr + (bestOffset << lBlkSize);
-        addresses.push_back(AddrPriority(prefetch_addr, 0));
-        DPRINTF(HWPrefetch, "Generated prefetch %#lx\n", prefetch_addr);
+        for (int d = 1; d <= degree; d++) {
```
Nit: in the paper he uses d as offset. I suggest renaming this variable to avoid confusing

Setu-Gupta replied:
Fixed in the latest commit

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -201,27 +207,34 @@ BOP::bestOffsetLearning(Addr x)
+    if ((bestScore >= scoreMax) || (round >= roundMax))
+    {
             round = 0;
+            if (bestScore > badScore)
```
Could you document why this is checked? It has been years since I reviewed this code/properly read the paper, so I don´t recall how it should work anymore.

Setu-Gupta replied:
Fixed in the latest commit

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -114,22 +111,24 @@ BOP::delayQueueEventWrapper()
-    return (hash1 ^ hash2) & (Addr)(rrEntries - 1);
+    Addr lrrEntries = floorLog2(rrEntries);
+    Addr lineaddr = addr >> lBlkSize;
+    Addr hash = lineaddr ^ (lineaddr >> (lrrEntries << way));
```
¨The RR table is accessed through a simple hash function.
For instance, for a 256-entry RR table, we XOR the 8 least
significant line address bits with the next 8 bits to obtain the
table index. For 12-bit tags, we skip the 8 least significant
line address bits and extract the next 12 bits.¨

- Here we use (64-8)-bit tags, right?
- The paper does not mention using the way here, and I don´t recall what is its used for. Could you document why?

Setu-Gupta replied:
Since the current version of BOP was not working as expected, i.e. it wasn't providing the expected performance on Spec workloads, I started looking for other implementations of BOP to see what could be the problem. While examining the prefetcher, I found that it was experiencing a lot of hash collisions while indexing the RR Table. To figure out if this was intended behavior or not, I started searching for other implementations of BOP online and I found the original code submitted by the author for DPC-2 [here](https://comparch-conf.gatech.edu/dpc2/final_program.html).

If you look at the source code, you'll two functions `rr_index_left` and `rr_index_right` which are used to generate the index which would be used to index the RR table. This logic is implemented in the `index` function in this PR. In the author's code you'll see that he multiplies the RRINDEX by 2 for indexing the right side, whereas he doesn't do the same for the left side. In this PR the `index` function is called with 0 for left, whereas it is called with 1 for right  (refer to the enum `RRWay` defined at line 57 of `bop.hh`). Therefore, if we right shift with the parameter `way`, then when `way` corresponds to left, we'll right shift with 0, and when it corresponds to right, we'll right shift by 1, i.e. multiply by 2. Thus, this code is equivalent to what the author submitted for DPC-2.

Setu-Gupta replied:
> Here we use (64-8)-bit tags, right?

We will use 8 bit tags if there are 256 entries in the RR Table. While this is the default size of the RR Table, it is not a constant. One can in fact change the size if they wish to do so.

> The paper does not mention using the way here, and I don´t recall what is its used for. Could you document why?

Since the reason why I made this change is not mentioned in the paper, should I still add a comment explaining the logic?

odanrc replied:
Yes, please add a comment explaining this logic! Your explanation was great, and will make it much easier to understand the code.

Setu-Gupta replied:
I've added a comment explaining the calculation

odanrc reviewed (changes requested):
I had to resort to the paper (I just skimmed across) to try to understand what is being fixed. My points of issue with the [Issue](https://github.com/gem5/gem5/issues/1402) were the use of ¨incorrectly¨ in bullet 2 without any description of what is the expected correct methodology. Similarly, in bullet 3 it would be nice to have a summary of how it was implemented, and how it was supposed to be implemented. Could you fill this information in the Issue so that the next review is more focused on the algorithm?

Commit:
mem: Fixed formatting and improved readability

Thanks to @ odanrc

On branch bop
Changes to be committed:
	modified:   src/mem/cache/prefetch/bop.cc
	modified:   src/mem/cache/prefetch/bop.hh

Co-authored-by: Abhishek Shailendra Singh <abs218@leigh.edu>
Change-Id: I59116dbd91a14dfd18336c8519e87ac68b3973e5

Setu-Gupta commented:
> I had to resort to the paper (I just skimmed across) to try to understand what is being fixed. My points of issue with the [Issue](https://github.com/gem5/gem5/issues/1402) were the use of ¨incorrectly¨ in bullet 2 without any description of what is the expected correct methodology. Similarly, in bullet 3 it would be nice to have a summary of how it was implemented, and how it was supposed to be implemented. Could you fill this information in the Issue so that the next review is more focused on the algorithm?

I've updated the issue. I hope now the reviewers would be able to understand the changes.

odanrc commented on src/mem/cache/prefetch/bop.cc:
```diff
@@ -201,27 +215,37 @@ BOP::bestOffsetLearning(Addr x)
             issuePrefetchRequests = true;
-        } else if (bestScore <= badScore) {
+        }
+        else {
```
According to the guidelines the else must be in the same line as the }

Setu-Gupta replied:
Fixed. By the way, is there a clang format file which outlines all the rules for Gem5 somewhere? I'm asking as right now I'm fixing the formatting manually because of which I sometimes miss some changes. With a clang-format file, all this work can be automated. Moreover, we can also add it to the commit hooks so that it automatically lints the code before committing

odanrc commented:
> > I had to resort to the paper (I just skimmed across) to try to understand what is being fixed. My points of issue with the [Issue](https://github.com/gem5/gem5/issues/1402) were the use of ¨incorrectly¨ in bullet 2 without any description of what is the expected correct methodology. Similarly, in bullet 3 it would be nice to have a summary of how it was implemented, and how it was supposed to be implemented. Could you fill this information in the Issue so that the next review is more focused on the algorithm?
> 
> I've updated the issue. I hope now the reviewers would be able to understand the changes.

That update was great! Thanks!

Commit:
mem: Fixed formatting and explained index calculation

On branch bop
Your branch is up to date with 'origin/bop'.
Changes to be committed:
	modified:   src/mem/cache/prefetch/bop.cc

Change-Id: I25bb8e656e7cfd09c87f8e59fda2304e87745710
Co-authored-by: Abhishek Shailendra Singh <abs218@leigh.edu>

powerjg commented:
>  By the way, is there a clang format file which outlines all the rules for Gem5 somewhere? I'm asking as right now I'm fixing the formatting manually because of which I sometimes miss some changes. With a clang-format file, all this work can be automated. Moreover, we can also add it to the commit hooks so that it automatically lints the code before committing

Great idea! https://github.com/orgs/gem5/discussions/333

odanrc reviewed (approved):
Thanks for the changes!
